### ISSUE TITLE
Make the search path for .coverage files configurable

### ISSUE THREAD

**Issue by aantron** [aantron/bisect_ppx#269]
Right now, the Bisect_ppx reporter checks if any `.coverage` files are specified on the command line. If not, it looks non-recursively in `.` and recursively `_build` (the reason for this scheme is to avoid listing build and dependency directories of package managers, like `./node_modules` and `./_opam`, since these can be very large).

A way should be added to specify the base path for `.coverage` file search.

The current code is:

https://github.com/aantron/bisect_ppx/blob/61399f61ea830425adaca039eb827447258f1365/src/report/report.ml#L291-L316

**Comment by aantron** [comment]
Since `.coverage` files can still be specified directly (after being found by, say, `find`), and since the underlying issue was that coverage files generated by a run inside esy could not be easily found, I'm closing this for now in favor of only doing #270, which is about automatically searching for `.coverage` files in esy.

**Comment by aantron** [comment]
From https://github.com/esy/esy/issues/1062#issuecomment-601194426:

> @aantron for reference, the genType tests consist on running on several projects. This means the .coverage files are spread across 4 different locations in nested directories.
I don't know how much of a corner case this is, but it might be worth adding an options to bisect-ppx-report, for those kinds of use cases.
For now I've simply copied them first.

@cristianoc Do you mean something like this issue, which is about adding some kind of flag to the reporter to tell it what directories to look in? If so, would you like the search from the given directories to be recursive or non-recursive? I assume recursive is a good default.

**Comment by cristianoc** [comment]
@aantron definitely recursive. But I would not want to look into, e.g. `node_modules`, which is likely to be in the same dir.
So an option where the parameter is `src` and recursive by default, should catch most use cases I think.

**Comment by cristianoc** [comment]
Btw, now that I can see the output -- it's great. Thanks a lot for this amazing tool.

**Comment by aantron** [comment]
Great :) I'll add something that can be used like `--search-path src/ --search-path blah/`. If specified, it will replace the current recursive search in `_build/default/` and non-recursive search in `.`. Also, if specified, the search will happen whether or not any `.coverage` files are manually passed on the command line (by the way, have you tried specifying all your `.coverage` files, after finding them using a bunch of `find` commands?).

**Comment by cristianoc** [comment]
@aantron to return the favor, here's the static "coverage" reported by the global dead code analyzer I'm working on, applied to the bisect_ppx project:

```
  Warning Dead Value
  src/common/bisect_common.mli 20:1
  try_finally is never used
  <-- line 20
  val try_finally : 'a -> ('a -> 'b) -> ('a -> unit) -> 'b [@@dead "try_finally"] 

  Warning Dead Value
  src/common/bisect_common.mli 25:1
  try_in_channel is never used
  <-- line 25
  val try_in_channel : bool -> string -> (in_channel -> 'a) -> 'a [@@dead "try_in_channel"] 

  Warning Dead Value
  src/common/bisect_common.mli 79:1
  table is never used
  <-- line 79
  val table : (string, int array * string) Hashtbl.t Lazy.t [@@dead "table"] 

  Warning Dead Value
  src/report/report_utils.ml 23:11
  Infix.+res is never used
  <-- line 23
        let res = (++) x max_int [@@dead "Infix.+res"]  in

  Warning Dead Value
  src/report/report_utils.ml 19:7
  Infix.-- is never used
  <-- line 19
          res [@@dead "Infix.--"] 

  Warning Dead Value
  src/report/report_utils.ml 43:7
  Infix.-| is never used
  <-- line 43
    let (-|) x y = zip (--) x y [@@dead "Infix.-|"] 

  Warning Dead Value
  src/report/report_utils.ml 66:11
  +spl is never used
  <-- line 66
      | [] -> (List.rev acc), [] [@@dead "+spl"]  in

  Warning Dead Value
  src/report/report_utils.ml 65:5
  split_after is never used
  <-- line 65
    spl n [] l [@@dead "split_after"] 

  Warning Dead Value
  src/report/report_utils.ml 109:5
  output_bytes is never used
  <-- line 109
      (fun channel -> Array.iter (output_byte channel) data) [@@dead "output_bytes"] 

  Warning Dead Value
  src/report/report_utils.ml 129:5
  sum is never used
  <-- line 129
  let sum = List.fold_left add (make ()) [@@dead "sum"] 

  Warning Dead Value
  src/report/report_utils.mli 16:3
  Infix.++ is never used
  <-- line 16
    val (++) : int -> int -> int [@@dead "Infix.++"] 

  Warning Dead Value
  src/report/report_utils.mli 21:3
  Infix.-- is never used
  <-- line 21
    val (--) : int -> int -> int [@@dead "Infix.--"] 

  Warning Dead Value
  src/report/report_utils.mli 32:3
  Infix.-| is never used
  <-- line 32
    val (-|) : int array -> int array -> int array [@@dead "Infix.-|"] 

  Warning Dead Value
  src/report/report_utils.mli 50:1
  split_after is never used
  <-- line 50
  val split_after : int -> ('a list) -> 'a list * 'a list [@@dead "split_after"] 

  Warning Dead Value
  src/report/report_utils.mli 70:1
  output_bytes is never used
  <-- line 70
  val output_bytes : int array -> string -> unit [@@dead "output_bytes"] 

  Warning Dead Value
  src/report/report_utils.mli 99:1
  sum is never used
  <-- line 99
  val sum : counts list -> counts [@@dead "sum"] 

  Warning Dead Value
  src/runtime/native/runtime.ml 96:5
  reset_counters is never used
  <-- line 96
      ) (Lazy.force Common.table) [@@dead "reset_counters"] 

  Warning Dead Value
  src/runtime/native/runtime.mli 50:1
  get_coverage_data is never used
  <-- line 50
  val get_coverage_data : unit -> string option [@@dead "get_coverage_data"] 

  Warning Dead Value
  src/runtime/native/runtime.mli 60:1
  dump_counters_exn is never used
  <-- line 60
  val dump_counters_exn : out_channel -> unit [@@dead "dump_counters_exn"] 

  Warning Dead Value
  src/runtime/native/runtime.mli 66:1
  reset_counters is never used
  <-- line 66
  val reset_counters : unit -> unit [@@dead "reset_counters"] 
```

If something is reprted dead only in the interface, it means it's never used from the outside (but could be used from the inside).

**Comment by aantron** [comment]
Nice! I saved it in https://github.com/aantron/bisect_ppx/issues/282, will likely go through it in the coming days, before the next release.

**Comment by aantron** [comment]
This is included in [2.1.0](https://github.com/aantron/bisect_ppx/releases/tag/2.1.0), which is already in npm, and should be in opam soon (https://github.com/ocaml/opam-repository/pull/16047).
